<a href="https://colab.research.google.com/github/dastef1984/russmann/blob/master/Pipeline(10_2024).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install recbole
!pip install ray
!pip install kmeans-pytorch

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 15.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.6/65.6 MB 14.9 MB/s eta 0:00:00


In [2]:
import os
from recbole.quick_start import run_recbole

# Path where models will be saved
checkpoint_dir = './saved_models/'
os.makedirs(checkpoint_dir, exist_ok=True)

# Dataset and Model Configuration
dataset_name = 'ml-100k'
data_path = 'dataset/'

# Define metrics for evaluation (only NDCG for now)
metrics = ['MRR']

# Check if the dataset path exists
if not os.path.exists(data_path):
    print(f"Dataset path {data_path} does not exist. Please ensure the dataset is available.")
else:
    print(f"Dataset {dataset_name} is ready.")


Dataset path dataset/ does not exist. Please ensure the dataset is available.


In [3]:
import os
import shutil

# Function to get configuration dictionary for each model
def get_config(model_name):
    config_dict = {
        'model': model_name,
        'dataset': dataset_name,
        'data_path': data_path,
        'epochs': 10,    # Modify if you want to adjust training duration
        'topk': 10,      # We'll evaluate Top-K items (for MRR)
        'metrics': metrics,
        'checkpoint_dir': checkpoint_dir,  # Define checkpoint directory
        'save_model': True,  # Ensure the model is saved
    }
    return config_dict

# Function to rename the model files after training
def rename_model_file(model_name):
    # Find the latest file saved in the checkpoint directory
    latest_file = max(
        [f for f in os.listdir(checkpoint_dir) if model_name in f],
        key=lambda x: os.path.getctime(os.path.join(checkpoint_dir, x))
    )

    # Define the new name for the model
    new_model_file = os.path.join(checkpoint_dir, f'{model_name}.pth')

    # Rename the file
    shutil.move(os.path.join(checkpoint_dir, latest_file), new_model_file)
    print(f"Model saved and renamed to {new_model_file}")

# Function to train and evaluate the model
def train_and_save_model(model_name):
    config_dict = get_config(model_name)
    print(f"Training and saving model: {model_name}...")

    # Train and save the model using RecBole
    result = run_recbole(config_dict=config_dict)

    # Rename the saved model to avoid long names
    rename_model_file(model_name)

# Train and save models with simplified filenames
train_and_save_model('BPR')
train_and_save_model('ItemKNN')
train_and_save_model('Pop')


Training and saving model: BPR...


/usr/local/lib/python3.10/dist-packages/recbole/data/dataset/dataset.py:648: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  feat[field].fillna(value=0, inplace=True)
/usr/local/lib/python3.10/dist-packages/recbole/data/dataset/dataset.py:650: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a c

Model saved and renamed to ./saved_models/BPR.pth
Training and saving model: ItemKNN...


/usr/local/lib/python3.10/dist-packages/recbole/data/dataset/dataset.py:648: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  feat[field].fillna(value=0, inplace=True)
/usr/local/lib/python3.10/dist-packages/recbole/data/dataset/dataset.py:650: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a c

Model saved and renamed to ./saved_models/ItemKNN.pth
Training and saving model: Pop...


/usr/local/lib/python3.10/dist-packages/recbole/data/dataset/dataset.py:648: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  feat[field].fillna(value=0, inplace=True)
/usr/local/lib/python3.10/dist-packages/recbole/data/dataset/dataset.py:650: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a c

Model saved and renamed to ./saved_models/Pop.pth


In [4]:
from recbole.quick_start import load_data_and_model

# Function to load the saved model and return model, dataset, and dataloader
def load_saved_model(model_name):
    """
    Load a saved model along with the dataset and dataloader.

    :param model_name: The name of the model to load (BPR, ItemKNN, Pop)
    :return: config, model, dataset, dataloader_train, dataloader_test
    """
    model_file = f'./saved_models/{model_name}.pth'  # Path to the saved model

    print(f"Loading saved model: {model_name} from {model_file}...")

    # Load the model, config, dataset, and dataloaders
    result = load_data_and_model(model_file)

    # Unpack and return the necessary components
    config, model, dataset, dataloader_train, dataloader_test, _ = result
    print(f"Model {model_name} loaded successfully.")

    return config, model, dataset, dataloader_train, dataloader_test

# Load all three models: BPR, ItemKNN, Pop
bpr_config, bpr_model, bpr_dataset, bpr_dataloader_train, bpr_dataloader_test = load_saved_model('BPR')
itemknn_config, itemknn_model, itemknn_dataset, itemknn_dataloader_train, itemknn_dataloader_test = load_saved_model('ItemKNN')
pop_config, pop_model, pop_dataset, pop_dataloader_train, pop_dataloader_test = load_saved_model('Pop')

Loading saved model: BPR from ./saved_models/BPR.pth...


/usr/local/lib/python3.10/dist-packages/recbole/quick_start/quick_start.py:250: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_file)
/usr/local/

Model BPR loaded successfully.
Loading saved model: ItemKNN from ./saved_models/ItemKNN.pth...


/usr/local/lib/python3.10/dist-packages/recbole/quick_start/quick_start.py:250: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_file)
/usr/local/

Model ItemKNN loaded successfully.
Loading saved model: Pop from ./saved_models/Pop.pth...


/usr/local/lib/python3.10/dist-packages/recbole/quick_start/quick_start.py:250: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_file)
/usr/local/

Model Pop loaded successfully.


In [5]:
import torch

# Function to generate top-k recommendations for a given model and dataloader
def get_topk_recommendations(model, dataloader, topk=10):
    """
    Generate top-k recommendations for a given model and dataloader.

    :param model: The trained model (e.g., BPR, ItemKNN, Pop)
    :param dataloader: The dataloader for the test set
    :param topk: The number of top recommendations to generate (default: 10)
    :return: List of top-k recommended items for each user
    """
    model.eval()  # Set model to evaluation mode
    topk_recommendations = []  # Placeholder to store top-k recommendations

    # Generate recommendations using the model and dataloader
    for data in dataloader:
        interaction = data[0]  # Get the interaction tensor (user, item interactions)
        scores = model.full_sort_predict(interaction)  # Predict scores for all items
        topk_items = torch.topk(scores, k=topk, dim=-1).indices  # Get top-k items
        topk_recommendations.append(topk_items.cpu().numpy())  # Append to the list

    return topk_recommendations

# Generate top-k recommendations for each model and store them in memory
print("Generating Top-K (10) Recommendations for each model...")

# BPR Model Recommendations
bpr_topk_recs = get_topk_recommendations(bpr_model, bpr_dataloader_test, topk=10)
print(f"Top-10 Recommendations for BPR Model:\n{bpr_topk_recs[:5]}")  # Print first 5 recommendations for inspection

# ItemKNN Model Recommendations
itemknn_topk_recs = get_topk_recommendations(itemknn_model, itemknn_dataloader_test, topk=10)
print(f"Top-10 Recommendations for ItemKNN Model:\n{itemknn_topk_recs[:5]}")  # Print first 5 recommendations for inspection

# Pop Model Recommendations
pop_topk_recs = get_topk_recommendations(pop_model, pop_dataloader_test, topk=10)
print(f"Top-10 Recommendations for Pop Model:\n{pop_topk_recs[:5]}")  # Print first 5 recommendations for inspection

Generating Top-K (10) Recommendations for each model...
Top-10 Recommendations for BPR Model:
[array([1736,  358,   53, 2041,   50, 1733, 2087, 1744,  290,  102]), array([1736, 2041, 1733, 2087, 1900, 1785, 1715, 1841, 1845, 1931]), array([1736, 2041, 1733, 1900, 1785, 2087, 1715, 1845, 1931, 1841]), array([1736, 2041,   53, 1733,  358,   50, 1785, 1900, 1715, 1840]), array([1736, 2041, 1733,  358, 1785,   53, 1900,   50, 1715,  102])]
Top-10 Recommendations for ItemKNN Model:
[array([1824, 1707, 2087, 1820, 2048, 1962, 2160, 1709, 1786, 1781]), array([1931, 1845, 1796, 1715, 1875, 1785, 1900, 1741, 1863, 2005]), array([1784, 1741, 1785, 1924, 1938, 1802, 2041, 1875, 1783, 2155]), array([1785, 1796, 1900, 1931, 1845, 2041, 1736, 1923, 1786, 1875]), array([ 102,  217,  113,  103, 1785,  248,  348,  322,  162,   54])]
Top-10 Recommendations for Pop Model:
[array([ 32,  53, 137, 192, 157, 210, 290, 358, 190,  78]), array([ 32,  53, 137, 192, 157, 210, 290, 358, 190,  78]), array([ 32,  53

In [6]:
import torch.nn.functional as F

def greedy_rerank(model, dataloader, dataset, topk=10, diversity_weight=0.5):
    """
    Greedy re-ranking to balance relevance (model's score) and diversity
    among the recommended items.

    :param model: The recommendation model (BPR, ItemKNN, or Pop)
    :param dataloader: The test dataloader for generating recommendations
    :param dataset: The dataset object to access item_num
    :param topk: Number of items to recommend after re-ranking (default: 10)
    :param diversity_weight: Weight controlling the trade-off between relevance and diversity (0 = no diversity, 1 = max diversity)
    :return: Re-ranked top-k recommendations
    """
    model.eval()  # Set the model to evaluation mode
    re_ranked_recommendations = []  # Placeholder for re-ranked recommendations
    num_items = dataset.item_num  # Number of items in the dataset

    for data in dataloader:
        interaction = data[0]  # Get interaction tensor
        scores = model.full_sort_predict(interaction).squeeze()  # Predict scores for all items and flatten
        initial_topk_items = torch.topk(scores, k=topk*2, dim=-1).indices  # Get initial 2x top-k items

        selected_items = []  # Placeholder for selected items after re-ranking

        # Add the first item (highest relevance score)
        selected_items.append(initial_topk_items[0].item())

        # Greedily add items to maximize diversity
        for _ in range(1, topk):
            best_item = None
            best_score = -float('inf')

            # Loop over the remaining items and select the best one based on a trade-off between score and diversity
            for item in initial_topk_items:
                if item.item() in selected_items:
                    continue  # Skip already selected items

                if item.item() >= num_items:  # Ensure item index is valid
                    continue  # Skip items that are out of bounds

                # Different logic for models without embeddings
                if hasattr(model, 'item_embedding'):  # BPR model with embeddings
                    # Calculate diversity: Use cosine similarity
                    diversity = 0
                    for selected_item in selected_items:
                        if selected_item >= num_items:  # Ensure selected_item index is valid
                            continue

                        sim = F.cosine_similarity(model.item_embedding.weight[item].unsqueeze(0),
                                                  model.item_embedding.weight[selected_item].unsqueeze(0)).item()
                        diversity += 1 - sim  # Higher difference = more diversity

                    diversity = diversity / len(selected_items)  # Average diversity for the current item
                else:
                    # For models like ItemKNN or Pop, we'll just consider the relevance scores
                    diversity = 0

                # Relevance score for the current item
                score = scores[item].item()

                # Combine relevance and diversity into a final score
                combined_score = (1 - diversity_weight) * score + diversity_weight * diversity

                # Select the item with the best combined score
                if combined_score > best_score:
                    best_score = combined_score
                    best_item = item.item()

            # Add the best item to the selected items
            if best_item is not None:
                selected_items.append(best_item)

        re_ranked_recommendations.append(selected_items)

    return re_ranked_recommendations


# Apply greedy re-ranking for each model's recommendations
print("Applying Greedy Re-ranking for each model...")

# BPR Re-ranked Recommendations (uses item embeddings)
bpr_re_ranked_recs = greedy_rerank(bpr_model, bpr_dataloader_test, bpr_dataset, topk=10, diversity_weight=0.5)
print(f"Re-ranked Top-10 Recommendations for BPR Model:\n{bpr_re_ranked_recs[:5]}")  # Print first 5 re-ranked recommendations

# ItemKNN Re-ranked Recommendations (without item embeddings)
itemknn_re_ranked_recs = greedy_rerank(itemknn_model, itemknn_dataloader_test, itemknn_dataset, topk=10, diversity_weight=0.5)
print(f"Re-ranked Top-10 Recommendations for ItemKNN Model:\n{itemknn_re_ranked_recs[:5]}")  # Print first 5 re-ranked recommendations

# Pop Re-ranked Recommendations (without item embeddings)
pop_re_ranked_recs = greedy_rerank(pop_model, pop_dataloader_test, pop_dataset, topk=10, diversity_weight=0.5)
print(f"Re-ranked Top-10 Recommendations for Pop Model:\n{pop_re_ranked_recs[:5]}")  # Print first 5 re-ranked recommendations

Applying Greedy Re-ranking for each model...
Re-ranked Top-10 Recommendations for BPR Model:
[[1736, 358, 53, 50, 290, 102, 158, 61, 190], [1736], [1736], [1736, 53, 358, 50, 158, 102, 61, 404, 217], [1736, 358, 53, 50, 102, 217, 32, 162]]
Re-ranked Top-10 Recommendations for ItemKNN Model:
[[1824], [1931], [1784], [1785], [102, 217, 113, 103, 248, 348, 322, 162, 54, 192]]
Re-ranked Top-10 Recommendations for Pop Model:
[[1820, 192, 290, 53, 32, 78, 210, 137, 190, 358], [1820, 192, 290, 53, 32, 78, 210, 137, 190, 358], [1820, 192, 290, 53, 32, 78, 210, 137, 190, 358], [1820, 192, 290, 53, 32, 78, 210, 137, 190, 358], [1820, 192, 290, 53, 32, 78, 210, 137, 190, 358]]


In [7]:
import numpy as np
import torch

# Reuse the NDCG calculation function you provided
def calculate_ndcg_at_k(topk_recs, ground_truth, k=10):
    def dcg_at_k(recommended_items, relevant_items, k):
        dcg = 0.0
        for i in range(min(k, len(recommended_items))):
            if recommended_items[i] in relevant_items:
                dcg += 1 / np.log2(i + 2)
        return dcg

    def idcg_at_k(relevant_items, k):
        idcg = 0.0
        for i in range(min(k, len(relevant_items))):
            idcg += 1 / np.log2(i + 2)
        return idcg

    total_ndcg = 0.0
    num_users = len(topk_recs)

    for user_idx in range(num_users):
        recommended_items = topk_recs[user_idx]
        relevant_items = ground_truth[user_idx]

        # Ensure both are lists or tensors
        if isinstance(recommended_items, torch.Tensor):
            recommended_items = recommended_items.tolist()
        if isinstance(relevant_items, torch.Tensor):
            relevant_items = relevant_items.tolist()
        if isinstance(recommended_items, int):
            recommended_items = [recommended_items]
        if isinstance(relevant_items, int):
            relevant_items = [relevant_items]

        # Calculate DCG@K for this user
        dcg = dcg_at_k(recommended_items, relevant_items, k)
        # Calculate IDCG@K for this user
        idcg = idcg_at_k(relevant_items, k)

        if idcg == 0:
            ndcg = 0.0
        else:
            ndcg = dcg / idcg

        total_ndcg += ndcg

    return total_ndcg / num_users

# Extract ground truth function
def extract_ground_truth_from_dataset(dataset, test=True):
    ground_truth = []
    data = dataset.inter_feat if test else dataset.train_data

    for i in range(len(data)):
        item_id = data['item_id'][i]
        ground_truth.append(item_id)

    return ground_truth

# Extract the ground truth interactions from the test dataset for each model
bpr_ground_truth = extract_ground_truth_from_dataset(bpr_dataset, test=True)
itemknn_ground_truth = extract_ground_truth_from_dataset(itemknn_dataset, test=True)
pop_ground_truth = extract_ground_truth_from_dataset(pop_dataset, test=True)

# Calculate NDCG for non-re-ranked recommendations
print("Evaluating non-re-ranked recommendations:")
ndcg_bpr = calculate_ndcg_at_k(bpr_topk_recs, bpr_ground_truth, k=10)
ndcg_itemknn = calculate_ndcg_at_k(itemknn_topk_recs, itemknn_ground_truth, k=10)
ndcg_pop = calculate_ndcg_at_k(pop_topk_recs, pop_ground_truth, k=10)

print(f"NDCG@10 for BPR: {ndcg_bpr}")
print(f"NDCG@10 for ItemKNN: {ndcg_itemknn}")
print(f"NDCG@10 for Pop: {ndcg_pop}")

# Calculate NDCG for re-ranked recommendations
print("\nEvaluating re-ranked recommendations:")
ndcg_bpr_reranked = calculate_ndcg_at_k(bpr_re_ranked_recs, bpr_ground_truth, k=10)
ndcg_itemknn_reranked = calculate_ndcg_at_k(itemknn_re_ranked_recs, itemknn_ground_truth, k=10)
ndcg_pop_reranked = calculate_ndcg_at_k(pop_re_ranked_recs, pop_ground_truth, k=10)

print(f"NDCG@10 for Re-ranked BPR: {ndcg_bpr_reranked}")
print(f"NDCG@10 for Re-ranked ItemKNN: {ndcg_itemknn_reranked}")
print(f"NDCG@10 for Re-ranked Pop: {ndcg_pop_reranked}")

Evaluating non-re-ranked recommendations:
NDCG@10 for BPR: 0.01211335827200395
NDCG@10 for ItemKNN: 0.004912738467097417
NDCG@10 for Pop: 0.028602174116840032

Evaluating re-ranked recommendations:
NDCG@10 for Re-ranked BPR: 0.014763416544407262
NDCG@10 for Re-ranked ItemKNN: 0.005555746321842406
NDCG@10 for Re-ranked Pop: 0.021414940828463416


In [8]:
import numpy as np

def calculate_average_popularity(topk_recs, item_popularity):
    total_popularity = 0
    num_users = len(topk_recs)

    for recs in topk_recs:
        # Sum the popularity of recommended items, skip if index is out of range
        total_popularity += sum(item_popularity[item] for item in recs if item < len(item_popularity))

    return total_popularity / (num_users * len(topk_recs[0]))


def calculate_gini_index(topk_recs, item_popularity):
    all_recs = [item for recs in topk_recs for item in recs if item < len(item_popularity)]
    item_freq = np.array([all_recs.count(i) for i in range(len(item_popularity))])
    item_freq = np.sort(item_freq)
    n = len(item_freq)
    index = np.arange(1, n + 1)
    gini = (2 * np.sum(index * item_freq) / np.sum(item_freq)) - (n + 1) / n
    return gini


# Function to calculate ItemCoverage
def calculate_item_coverage(topk_recs, total_items):
    unique_items = set(item for recs in topk_recs for item in recs)
    coverage = len(unique_items) / total_items
    return coverage

# Function to calculate ShannonEntropy
def calculate_shannon_entropy(topk_recs):
    all_recs = [item for recs in topk_recs for item in recs]
    item_freq = np.array([all_recs.count(i) for i in set(all_recs)])
    prob = item_freq / item_freq.sum()
    entropy = -np.sum(prob * np.log2(prob))
    return entropy

def calculate_tail_percentage(topk_recs, item_popularity, tail_threshold=0.2):
    num_items = len(item_popularity)
    threshold = int(tail_threshold * num_items)  # Items in the tail are the least popular ones
    tail_items = set(np.argsort(item_popularity)[:threshold])

    num_tail_items = sum(item in tail_items for recs in topk_recs for item in recs if item < len(item_popularity))
    total_recommendations = len(topk_recs) * len(topk_recs[0])

    return num_tail_items / total_recommendations


In [9]:
# Calculate item popularity from the dataset
def calculate_item_popularity(dataset):
    item_popularity = {}

    # Iterate through all interactions in the dataset
    for item_id in dataset['item_id']:
        if item_id in item_popularity:
            item_popularity[item_id] += 1
        else:
            item_popularity[item_id] = 1

    # Convert to a list where index is item_id and value is the popularity
    total_items = dataset.item_num  # Ensure this covers all items in the dataset
    item_popularity_list = [item_popularity.get(i, 0) for i in range(total_items)]

    return item_popularity_list


# Example usage after loading the dataset
item_popularity = calculate_item_popularity(bpr_dataset)  # Calculate item popularity for the BPR dataset

In [10]:
# Calculate the new metrics for non-re-ranked recommendations
average_popularity_bpr = calculate_average_popularity(bpr_topk_recs, item_popularity)
gini_index_bpr = calculate_gini_index(bpr_topk_recs, item_popularity)
item_coverage_bpr = calculate_item_coverage(bpr_topk_recs, bpr_dataset.item_num)
shannon_entropy_bpr = calculate_shannon_entropy(bpr_topk_recs)
tail_percentage_bpr = calculate_tail_percentage(bpr_topk_recs, item_popularity)

# Print the results for BPR model
print(f"AveragePopularity for BPR: {average_popularity_bpr}")
print(f"GiniIndex for BPR: {gini_index_bpr}")
print(f"ItemCoverage for BPR: {item_coverage_bpr}")
print(f"ShannonEntropy for BPR: {shannon_entropy_bpr}")
print(f"TailPercentage for BPR: {tail_percentage_bpr}")

# Calculate metrics for re-ranked recommendations
average_popularity_bpr_reranked = calculate_average_popularity(bpr_re_ranked_recs, item_popularity)
gini_index_bpr_reranked = calculate_gini_index(bpr_re_ranked_recs, item_popularity)
item_coverage_bpr_reranked = calculate_item_coverage(bpr_re_ranked_recs, bpr_dataset.item_num)
shannon_entropy_bpr_reranked = calculate_shannon_entropy(bpr_re_ranked_recs)
tail_percentage_bpr_reranked = calculate_tail_percentage(bpr_re_ranked_recs, item_popularity)

# Print results for re-ranked BPR model
print(f"AveragePopularity for Re-ranked BPR: {average_popularity_bpr_reranked}")
print(f"GiniIndex for Re-ranked BPR: {gini_index_bpr_reranked}")
print(f"ItemCoverage for Re-ranked BPR: {item_coverage_bpr_reranked}")
print(f"ShannonEntropy for Re-ranked BPR: {shannon_entropy_bpr_reranked}")
print(f"TailPercentage for Re-ranked BPR: {tail_percentage_bpr_reranked}")

AveragePopularity for BPR: 0.0
GiniIndex for BPR: 3354.390600791489
ItemCoverage for BPR: 0.0285204991087344
ShannonEntropy for BPR: 4.88347990175488
TailPercentage for BPR: 0.0
AveragePopularity for Re-ranked BPR: 0.0
GiniIndex for Re-ranked BPR: 3353.7849043123615
ItemCoverage for Re-ranked BPR: 0.019013666072489603
ShannonEntropy for Re-ranked BPR: 4.129052069247205
TailPercentage for Re-ranked BPR: 0.0


In [11]:
def relevance_rerank(model, dataloader, dataset, topk=10):
    """
    Relevance re-ranking based purely on the model's predicted scores.

    :param model: The recommendation model (BPR, ItemKNN, or Pop)
    :param dataloader: The test dataloader for generating recommendations
    :param dataset: The dataset object to access item_num
    :param topk: Number of items to recommend after re-ranking (default: 10)
    :return: Re-ranked top-k recommendations
    """
    model.eval()  # Set the model to evaluation mode
    re_ranked_recommendations = []  # Placeholder for re-ranked recommendations

    for data in dataloader:
        interaction = data[0]  # Get interaction tensor
        scores = model.full_sort_predict(interaction).squeeze()  # Predict scores for all items and flatten
        topk_items = torch.topk(scores, k=topk, dim=-1).indices  # Get the top-k items based purely on relevance

        re_ranked_recommendations.append(topk_items.cpu().numpy())  # Append to the re-ranked list

    return re_ranked_recommendations

In [12]:
# Apply relevance-based re-ranking for each model
print("Applying Relevance-Based Re-ranking for each model...")

# BPR Relevance-Based Re-ranked Recommendations
bpr_relevance_re_ranked_recs = relevance_rerank(bpr_model, bpr_dataloader_test, bpr_dataset, topk=10)
print(f"Relevance-Based Re-ranked Top-10 Recommendations for BPR Model:\n{bpr_relevance_re_ranked_recs[:5]}")

# ItemKNN Relevance-Based Re-ranked Recommendations
itemknn_relevance_re_ranked_recs = relevance_rerank(itemknn_model, itemknn_dataloader_test, itemknn_dataset, topk=10)
print(f"Relevance-Based Re-ranked Top-10 Recommendations for ItemKNN Model:\n{itemknn_relevance_re_ranked_recs[:5]}")

# Pop Relevance-Based Re-ranked Recommendations
pop_relevance_re_ranked_recs = relevance_rerank(pop_model, pop_dataloader_test, pop_dataset, topk=10)
print(f"Relevance-Based Re-ranked Top-10 Recommendations for Pop Model:\n{pop_relevance_re_ranked_recs[:5]}")

Applying Relevance-Based Re-ranking for each model...
Relevance-Based Re-ranked Top-10 Recommendations for BPR Model:
[array([1736,  358,   53, 2041,   50, 1733, 2087, 1744,  290,  102]), array([1736, 2041, 1733, 2087, 1900, 1785, 1715, 1841, 1845, 1931]), array([1736, 2041, 1733, 1900, 1785, 2087, 1715, 1845, 1931, 1841]), array([1736, 2041,   53, 1733,  358,   50, 1785, 1900, 1715, 1840]), array([1736, 2041, 1733,  358, 1785,   53, 1900,   50, 1715,  102])]
Relevance-Based Re-ranked Top-10 Recommendations for ItemKNN Model:
[array([1824, 1707, 2087, 1820, 2048, 1962, 2160, 1709, 1786, 1781]), array([1931, 1845, 1796, 1715, 1875, 1785, 1900, 1741, 1863, 2005]), array([1784, 1741, 1785, 1924, 1938, 1802, 2041, 1875, 1783, 2155]), array([1785, 1796, 1900, 1931, 1845, 2041, 1736, 1923, 1786, 1875]), array([ 102,  217,  113,  103, 1785,  248,  348,  322,  162,   54])]
Relevance-Based Re-ranked Top-10 Recommendations for Pop Model:
[array([ 32,  53, 137, 192, 157, 210, 290, 358, 190,  78])

In [13]:
# Evaluate relevance-based re-ranked recommendations using the same metrics

# NDCG evaluation
print("\nEvaluating Relevance-Based Re-ranked recommendations:")
ndcg_bpr_relevance_reranked = calculate_ndcg_at_k(bpr_relevance_re_ranked_recs, bpr_ground_truth, k=10)
ndcg_itemknn_relevance_reranked = calculate_ndcg_at_k(itemknn_relevance_re_ranked_recs, itemknn_ground_truth, k=10)
ndcg_pop_relevance_reranked = calculate_ndcg_at_k(pop_relevance_re_ranked_recs, pop_ground_truth, k=10)

print(f"NDCG@10 for Relevance-Based Re-ranked BPR: {ndcg_bpr_relevance_reranked}")
print(f"NDCG@10 for Relevance-Based Re-ranked ItemKNN: {ndcg_itemknn_relevance_reranked}")
print(f"NDCG@10 for Relevance-Based Re-ranked Pop: {ndcg_pop_relevance_reranked}")

# Calculate the other metrics (e.g., Average Popularity, Gini Index, etc.)
average_popularity_bpr_relevance_reranked = calculate_average_popularity(bpr_relevance_re_ranked_recs, item_popularity)
gini_index_bpr_relevance_reranked = calculate_gini_index(bpr_relevance_re_ranked_recs, item_popularity)
item_coverage_bpr_relevance_reranked = calculate_item_coverage(bpr_relevance_re_ranked_recs, bpr_dataset.item_num)
shannon_entropy_bpr_relevance_reranked = calculate_shannon_entropy(bpr_relevance_re_ranked_recs)
tail_percentage_bpr_relevance_reranked = calculate_tail_percentage(bpr_relevance_re_ranked_recs, item_popularity)

# Print results for relevance-based re-ranked BPR model
print(f"AveragePopularity for Relevance-Based Re-ranked BPR: {average_popularity_bpr_relevance_reranked}")
print(f"GiniIndex for Relevance-Based Re-ranked BPR: {gini_index_bpr_relevance_reranked}")
print(f"ItemCoverage for Relevance-Based Re-ranked BPR: {item_coverage_bpr_relevance_reranked}")
print(f"ShannonEntropy for Relevance-Based Re-ranked BPR: {shannon_entropy_bpr_relevance_reranked}")
print(f"TailPercentage for Relevance-Based Re-ranked BPR: {tail_percentage_bpr_relevance_reranked}")


Evaluating Relevance-Based Re-ranked recommendations:
NDCG@10 for Relevance-Based Re-ranked BPR: 0.01211335827200395
NDCG@10 for Relevance-Based Re-ranked ItemKNN: 0.004912738467097417
NDCG@10 for Relevance-Based Re-ranked Pop: 0.028602174116840032
AveragePopularity for Relevance-Based Re-ranked BPR: 0.0
GiniIndex for Relevance-Based Re-ranked BPR: 3354.390600791489
ItemCoverage for Relevance-Based Re-ranked BPR: 0.0285204991087344
ShannonEntropy for Relevance-Based Re-ranked BPR: 4.88347990175488
TailPercentage for Relevance-Based Re-ranked BPR: 0.0


### Relevance-Based Re-ranking Results Overview

Let’s break down the results you’ve obtained for the **Relevance-Based Re-ranking** method and compare them with the previous **Greedy Re-ranking** and **original** recommendations:

### NDCG@10 Comparison:
- **Relevance-Based Re-ranked BPR:** 0.0121
- **Greedy Re-ranked BPR:** 0.0148
- **Original BPR:** 0.0121

**Observation:** The NDCG for the relevance-based re-ranking is identical to the original BPR recommendations, which is expected since this method prioritizes relevance without considering diversity. However, the greedy re-ranking slightly improved NDCG by incorporating diversity in addition to relevance.

### Average Popularity Comparison:
- **Relevance-Based Re-ranked BPR:** 0.0
- **Greedy Re-ranked BPR:** 0.0
- **Original BPR:** 0.0

**Observation:** The average popularity for all BPR models is 0, which likely indicates that the items recommended are very unpopular or niche, possibly due to a low number of interactions. This result remains the same across all methods.

### Gini Index Comparison:
- **Relevance-Based Re-ranked BPR:** 3354.39
- **Greedy Re-ranked BPR:** 3353.78
- **Original BPR:** 3354.39

**Observation:** The Gini Index, which measures inequality in recommendations, remains almost identical between relevance-based and original BPR, as expected. The greedy re-ranking slightly lowered it, which is good as it suggests more diversity.

### Item Coverage Comparison:
- **Relevance-Based Re-ranked BPR:** 0.0285
- **Greedy Re-ranked BPR:** 0.0190
- **Original BPR:** 0.0285

**Observation:** The relevance-based re-ranking provides the same item coverage as the original recommendations, while the greedy re-ranking reduced item coverage, indicating that greedy re-ranking may be focusing on a narrower set of items to optimize for diversity.

### Shannon Entropy Comparison:
- **Relevance-Based Re-ranked BPR:** 4.88
- **Greedy Re-ranked BPR:** 4.12
- **Original BPR:** 4.88

**Observation:** Relevance-based re-ranking has higher Shannon Entropy (similar to the original BPR), which suggests a more even distribution of recommended items. The greedy re-ranking reduced entropy, likely due to the focus on balancing diversity with relevance.

### Tail Percentage Comparison:
- **Relevance-Based Re-ranked BPR:** 0.0
- **Greedy Re-ranked BPR:** 0.0
- **Original BPR:** 0.0

**Observation:** Both the original and re-ranked recommendations (relevance-based and greedy) do not include any items from the tail, indicating that none of the recommended items are from the least popular portion of the catalog.

### Conclusion:
- **Relevance-Based Re-ranking** gives results that are almost identical to the **original** BPR recommendations because it doesn't add any additional constraints like diversity. It performs similarly on NDCG and other metrics like coverage and entropy.
- **Greedy Re-ranking**, however, introduces slight improvements in NDCG at the cost of reduced item coverage and entropy. This method likely performs better when diversity is valued alongside relevance.